packages/core/modules-sdk/src/remote-query/query.ts

```diff
@@ -259,17 +261,98 @@ export class Query {
   }
 }
 
+function extractCacheOptions(option: string) {
+  return function extractKey(args: any[]) {
+    return args[1]?.cache?.[option]
+  }
+}
+
+function isCacheEnabled(args: any[]) {
+  const isEnabled = extractCacheOptions("enable")(args)
+  if (isEnabled === false) {
+    return false
+  }
+
+  return (
+    extractCacheOptions("key")(args) ||
+    extractCacheOptions("ttl")(args) ||
+    extractCacheOptions("tags")(args) ||
+    extractCacheOptions("autoInvalidate")(args) ||
+    extractCacheOptions("providers")(args)
+  )
+}
+
 /**
  * API wrapper around the remoteQuery with backward compatibility support
  * @param remoteQuery
  */
 export function createQuery({
   remoteQuery,
   indexModule,
+  container,
 }: {
   remoteQuery: RemoteQuery
   indexModule: IIndexService
+  container: MedusaContainer
 }) {
+  Query.prototype.graph = Cached<Query, "graph">({
+    enable: isCacheEnabled,
+    key: async (args, cachingModule) => {
+      const key = extractCacheOptions("key")(args)
+      if (key) {
+        return key
+      }
+
+      const queryOptions = args[0]
+      const remoteJoinerOptions = args[1] ?? {}
+      const { initialData, cache, ...restOptions } = remoteJoinerOptions
+
+      const keyInput = {
+        queryOptions,
+        options: restOptions,
+      }
+      return await cachingModule.computeKey(keyInput)
+    },
+    ttl: extractCacheOptions("ttl"),
+    tags: extractCacheOptions("tags"),
+    autoInvalidate: extractCacheOptions("autoInvalidate"),
+    providers: extractCacheOptions("providers"),
+    container,
+  })(
+    Query.prototype,
+    "graph",
+    Object.getOwnPropertyDescriptor(Query.prototype, "graph")!
+  ).value
+
+  Query.prototype.index = Cached<Query, "index">({
+    enable: isCacheEnabled,
+    key: async (args, cachingModule) => {
+      const key = extractCacheOptions("key")(args)
+      if (key) {
+        return key
+      }
+
+      const queryOptions = args[0]
+      const remoteJoinerOptions = args[1] ?? {}
+      const { initialData, cache, ...restOptions } = remoteJoinerOptions
+
+      const keyInput = {
+        queryOptions,
+        options: restOptions,
+      }
+      return await cachingModule.computeKey(keyInput)
+    },
+    ttl: extractCacheOptions("ttl"),
+    tags: extractCacheOptions("tags"),
+    autoInvalidate: extractCacheOptions("autoInvalidate"),
+    providers: extractCacheOptions("providers"),
+    container,
+  })(
+    Query.prototype,
+    "index",
+    Object.getOwnPropertyDescriptor(Query.prototype, "index")!
+  ).value
```

Comment:
Self-review: I ll do some cleanup around here. Pass container to Query and apply decorator to the method directly